beacon-light-client/plonky2/input_fetchers/balance_verification/deposits_accumulator/lib/pubkey_commitment_mapper_scheduler.ts

```diff
@@ -0,0 +1,230 @@
+import { sleep } from '@dendreth/utils/ts-utils/common-utils';
+import { Contract, ethers } from 'ethers';
+import { fetchEventsAsyncCB } from './event_fetcher';
+import { ChainableCommander } from 'ioredis';
+
+const sharedPrefix = `pubkey_commitment_mapper`;
+
+const pubkeyProcessingQueueKey = `${sharedPrefix}:processing_queue`;
+const pubkeysKey = `${sharedPrefix}:pubkeys`;
+const currentlyComputedPubkeyMappingKey = `${sharedPrefix}:currently_computed_pubkey_mapping`;
+const lastLoggedBlockKey = `${sharedPrefix}:last_logged_block`;
+
+const depositEventName = 'Deposited';
+
+interface SchedulerContext {
+  redis: Redis;
+  ethJsonRPC: ethers.providers.JsonRpcProvider;
+  contract: Contract;
+  protocol: string;
+}
+
+interface SchedulerContextConfig {
+  redisHost: string;
+  redisPort: number;
+  ethJsonRPCProviderURL: string;
+  contractAddress: string;
+  contractAbi: any;
+  protocol: string;
+}
+
+export function createSchedulerContext(
+  config: SchedulerContextConfig,
+): SchedulerContext {
+  const redis = new Redis(config.redisHost, config.redisPort);
+
+  const ethJsonRPC = new ethers.providers.JsonRpcProvider(
+    config.ethJsonRPCProviderURL,
+  );
+
+  const contract = new Contract(
+    config.contractAddress,
+    config.contractAbi,
+    ethJsonRPC,
+  );
+
+  return {
+    redis,
+    ethJsonRPC,
+    contract,
+    protocol: config.protocol,
+  };
+}
+
+export async function destroySchedulerContext(
+  ctx: SchedulerContext,
+): Promise<void> {
+  await ctx.redis.quit();
+}
+
+export async function pollEvents(
+  ctx: SchedulerContext,
+  timeout: number = 5000,
+): Promise<void> {
+  const lastLoggedBlock = await ctx.redis.client.get(
+    `${ctx.protocol}:${lastLoggedBlockKey}`,
+  );
+  if (lastLoggedBlock === null) {
+    console.log('Error: Tree not initialized, use --rebuild');
+    await destroySchedulerContext(ctx);
+    process.exit(1);
+  }
+
+  while (true) {
+    const lastLoggedBlock = await getLastLoggedBlock(ctx);
+    const headBlockNumber = await ctx.ethJsonRPC.getBlockNumber();
+
+    await fetchEventsAsyncCB(
+      ctx.contract,
+      depositEventName,
+      lastLoggedBlock + 1,
+      headBlockNumber,
+      async event => {
+        const args = event.args!;
+        const pubkey = formatHex(args[0]);
+        await updateDepositState(ctx, pubkey, event.blockNumber);
+      },
+    );
+
+    await setLastLoggedBlock(ctx, headBlockNumber);
+    await sleep(timeout);
+  }
+}
+
+export async function rebuildCommitmentMapperTree(
+  ctx: SchedulerContext,
+): Promise<void> {
+  const contractDeploymentBlockNumber =
+    await queryContractDeploymentBlockNumber(ctx);
+
+  if (contractDeploymentBlockNumber === null) {
+    console.log('Error: Invalid contract address');
+    await destroySchedulerContext(ctx);
+    process.exit(1);
+  }
+
+  await purgePubkeyCommitmentMapperData(ctx);
+
+  await setCurrentlyComputedPubkeyMapping(ctx, 0);
+  await setLastLoggedBlock(ctx, contractDeploymentBlockNumber - 1);
```

Comment:
Why not use pipes here as well?

Reply:
It's not necessary as this is done once and the worker is not expected to be running at this time